subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy

```diff
@@ -407,6 +407,43 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec imp
         executedAndNotSkipped(":produceInBuild", ":showSources")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/16980")
+    def "missing dependency detection takes ** excludes for non-existing files into account"() {
+        file("src/main/java/my/JavaClass.java").text = """
+            package my;
+
+            public class JavaClass {}
+        """
+
+        buildFile """
+            task fooReport {
+                inputs.files(fileTree(projectDir) { include("**/*.foo")})
+                def reportPath = file("\${buildDir}/fooReport.txt")
+                outputs.file(reportPath)
+                doLast {
+                    reportPath.text = "foo"
+                }
+            }
+            task barReport {
+                inputs.files(fileTree(projectDir) { include("**/*.bar")})
+                def reportPath = file("\${buildDir}/barReport.txt")
+                outputs.file(reportPath)
+                doLast {
+                    reportPath.text = "bar"
+                }
+            }
+
+            task allReports {
+                dependsOn(fooReport, barReport)
+            }
+        """
```

Comment:
To me this example is less understandable than what the user uploaded. After some investigation now I understand how these inputs produce a similar situation, but I think we can do something where the problem is more obvious to the reader.

Reply:
This PR is actually fixing the problem from the second comment on the issue. I extracted the comment to its own issue: #17561